### PR TITLE
NxTooltip RTL Tests - RSC-1227

### PR DESCRIPTION
**Pull request by rpokorny**
https://issues.sonatype.org/browse/RSC-1227

### CODE REVIEW

File: lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx [hunk 1]
```diff
@@ -4,113 +4,278 @@
...
 
-    expect(component.children()).toHaveProp('title', title);
+    expect(tooltip).toBeInTheDocument();
+    expect(tooltip).toHaveTextContent('tip');
   });
 
-  it('passes "nx-tooltip" as the `tooltip` property on the `classes` prop of the Tooltip', function() {
-    expect(getTooltip().prop('classes').tooltip).toBe('nx-tooltip');
-  });
+  it('calls onOpen when the tooltip renders by hover but not due to the open prop', async function() {
```
Comment (EmilyBermudez88) [inline lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx:73]: When I tested the second quickRender function with different props (ie. `quickRender({ onOpen, className: 'foo', placement: 'bottom'})` ), and used `await user.hover(child)` for the second time, the test still came back as having `onOpen` being called one time. Perhaps the `open` prop set to `true` should be tested first to show `onOpen` wasn't called, and then use `quickRender({onOpen})` & `await user.hover(child)` to assert `onOpen` has indeed been called.

File: lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx [hunk 2]
```diff
@@ -4,113 +4,278 @@
...
+      expect(screen.queryByRole('tooltip')).not.toBeInTheDocument();
+
+      await user.hover(child);
+
+      await expect(screen.findByRole('tooltip')).rejects.toBeTruthy();
+    });
+
+    it('does not affect the accessible name or description of the child', function() {
+      const children = <button aria-describedby="descriptionId">Foo</button>,
+          view = quickRender({ children }),
```
Comment (EmilyBermudez88) [inline lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx:122]:
```suggestion
          view = quickRender({ children, title: '' }),
```

File: lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx [hunk 3]
```diff
@@ -4,113 +4,278 @@
...
+          await user.hover(button);
+          await waitFor(() => expect(button).toHaveAccessibleDescription('tiptip'));
+          expect(button).not.toHaveAttribute('title');
+        }
+    );
+
+    it('does not override the accessible description if it is already set',
+        async function() {
+          const user = userEvent.setup(),
+              childrenWithOwnDescription = <button aria-describedby="descriptionId">Foo</button>,
```
Comment (EmilyBermudez88) [inline lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx:271]: need the JSX title prop added here

File: lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx (hunk 2 above)
Comment (rpokorny) [reply lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx]: ea70cf4a2

File: lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx (hunk 3 above)
Comment (rpokorny) [reply lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx:272]: > If it's under the 'when title is JSX' describe, I think you need the JSX title prop added here

ea70cf4a2

File: lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx (hunk 1 above)
Comment (EmilyBermudez88) [reply lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx:73]: >When I change the code as follows, it passes

Thanks for writing out a solution. I called `hover` and `waitFor`, but completely forgot to assign a variable to the second `quickRender` (just called `quickRender({ onOpen, placement: 'bottom'}), which of course caused things to fail. Thanks!

**Review by EmilyBermudez88 — APPROVED** [review]

File: lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx [hunk 4]
```diff
@@ -4,113 +4,317 @@
...
+              viewWithOwnDescription = quickRender({ children: childrenWithOwnDescription }),
+              buttonWithOwnDescription = viewWithOwnDescription.getByRole('button');
+
+          await user.hover(buttonWithOwnDescription);
+          expect(buttonWithOwnDescription).toHaveAccessibleDescription('Description Text');
+          expect(buttonWithOwnDescription).not.toHaveAttribute('title');
+
+          // Ensure that nothing changes after waiting for tooltip to initialize
+          await expect(waitFor(() => expect(buttonWithOwnDescription).not.toHaveAccessibleDescription('Description Text')))
```
Comment (EmilyBermudez88) [inline lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx:215]: lintr error here with too many characters on line 215

File: lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx [hunk 5]
```diff
@@ -117,13 +119,19 @@ describe('NxTooltip', function() {
...
+    it('does not affect the accessible name or description of the child', async function() {
       const children = <button aria-describedby="descriptionId">Foo</button>,
           view = quickRender({ title: '', children }),
           button = view.getByRole('button');
 
       expect(button).toHaveAccessibleName('Foo');
       expect(button).toHaveAccessibleDescription('Description Text');
+
+      // Ensure that nothing changes after waiting for tooltip to initialize
+      await expect(waitFor(() => expect(button).not.toHaveAccessibleName('Foo')))
```
[the remark this answers is not in the document]
Comment (rpokorny) [reply lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx:131]: > For clarification, why do we need the test here? If the title is an empty string, wouldn't the tooltip fail to render on the page?

The issue is that the tooltips initialize asynchronously, so any test that's trying to test that they _did not_ change something really needs to ensure not just that nothing changed immediately, but that nothing changed after waiting a bit for the tooltip to initialize as well.

[the remark this answers is not in the document]
Comment (EmilyBermudez88) [reply lib/src/components/NxTooltip/__tests__/NxTooltip.test.tsx:131]: > The issue is that the tooltips initialize asynchronously, so any test that's trying to test that they _did not_ change something really needs to ensure not just that nothing changed immediately, but that nothing changed after waiting a bit for the tooltip to initialize as well.

Makes sense for the situation here, thanks!

**Review by EmilyBermudez88 — APPROVED** [review]
Once the eslint errors are fixed, looks good!
